python/cuml/dask/naive_bayes/naive_bayes.py

```diff
@@ -232,7 +236,48 @@ def predict(self, X, delayed=True, parallelism=5):
         dask.Array containing class predictions
         """
 
-        return self._predict(X, delayed, parallelism)
+        gpu_futures = self.client_.sync(extract_arr_partitions, X)
+        x_worker_parts = workers_to_parts(gpu_futures)
+
+        key = uuid1()
+        futures = [(wf[0],
+                    self.client_.submit(MultinomialNB._get_size,
+                                        wf[1],
+                                        workers=[wf[0]],
+                                        key="%s-%s" % (key, idx)))
+                   for idx, wf in enumerate(gpu_futures)]
+
+        sizes = self.client_.compute([x[1] for x in futures], sync=True)
+
+        models = self.client_.scatter(self.local_model,
```

Comment:
Do we need this scatter in advance? Wouldn't the predict below it automatically scatter the local_model as needed? I guess this is the optimization to reduce the amount of transfer when you have multiple partitions per worker, which is fair...

Reply:
There's a few different points to consider here, so I'll address them separately:

1.  Calling `client.submit(self.local_model)` will cause an implicit transfer of the model to the workers and this is okay for smaller models, but specifically the number of parameters for Naive Bayes are on the order  of `n_terms * n_classes`. I didn't find a whole lot of performance difference between doing this and using client.submit, but the warning that's printed is undesirable.   

2. In the completely lazy path, a scatter is done implicitly in the task graph. I have opened https://github.com/rapidsai/cuml/issues/1834 to refactor Naive Bayes like the other algorithms- to extend cuml.Base, use the lazy prediction mixing, etc... The main optimization here is that rather than scattering the objects separately as I was doing before, they can be scattered in one call. That was an oversight on my part.

3. An example of the implicit scatter in the fully-lazy path is here: https://github.com/rapidsai/cuml/blob/branch-0.13/python/cuml/dask/common/base.py#L89. Since `model` is a delayed object, it will be implicitly scattered out to the workers and Dask guarantees only one instance will be made per worker.

Reply:
Also to add to this- I believe one of the reasons the use of scatter here didn't demonstrate much of a performance difference is because the bulk of the data is stored in GPU and so there really isn't much difference in the serialization / transfer underneath.